### PR TITLE
interfaces/apparmor: add chopTree

### PR DESCRIPTION
**Pull request by zyga**
This patch adds a helper function for chopping a path into a pair of
apparmor path expressions. The left expression describes directories
we assume exist, the right expression describes directories and files
we assume may need to be created.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: interfaces/apparmor/spec.go [hunk 1]
```diff
@@ -173,6 +173,84 @@ func isProbablyPresent(path string) bool {
...
+//
+// Those rules allow the code to open (to traverse securely) /, then foo, then
+// bar and then create _either_ 1) any file in /foo/bar/ or 2) any directory in
+// /foo/bar/ or 3) the file froz in /foo/bar/.
+//
+// Those rules are useful for constructing the apparmor profile for a writable
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz) assuming that part of that directory already exists (e.g.
+// /foo/) but may need to be created earlier (e.g. in /foo/bar/) because
+// /foo/bar/ is a read-only file system.
```
Comment (chipaca) [inline interfaces/apparmor/spec.go]: May I suggest (given that it took me chatting with you on IRC for a while to get it) that you:
* bump the example to have two directories on either side of the split
* mention why the *s are needed

the second point could be adding something like “So a tmpfs is created and mounted over /foo/bar, and then everything the outer /foo/bar is created inside and bind-mounted over (this the need of the * rules)” to the last paragraph here. (it's likely you'll be able to come up with a clearer explanation than my botch job given how long you've been thinking about this problem).

File: interfaces/apparmor/spec.go [hunk 2]
```diff
@@ -173,6 +173,84 @@ func isProbablyPresent(path string) bool {
...
+	var left, right bytes.Buffer
+	parts := strings.Split(path, "/")
+	isAbs := strings.HasPrefix(path, "/")
+	isDir := strings.HasSuffix(path, "/")
+
+	if isAbs {
+		fmt.Fprintf(&left, "/")
+		fmt.Fprintf(&right, "/")
+	}
+	cnt := 0
```
Comment (chipaca) [inline interfaces/apparmor/spec.go]: *giggle*

File: interfaces/apparmor/spec.go [hunk 3]
```diff
@@ -173,6 +173,84 @@ func isProbablyPresent(path string) bool {
...
+// Those rules allow the code to open (to traverse securely) /, then foo, then
+// bar and then create _either_ 1) any file in /foo/bar/ or 2) any directory in
+// /foo/bar/ or 3) the file froz in /foo/bar/.
+//
+// Those rules are useful for constructing the apparmor profile for a writable
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz) assuming that part of that directory already exists (e.g.
+// /foo/) but may need to be created earlier (e.g. in /foo/bar/) because
+// /foo/bar/ is a read-only file system.
+func chopTree(path string, assumedPrefixDepth int) (string, string) {
```
Comment (chipaca) [inline interfaces/apparmor/spec.go]: how performance-sensitive is this code? I'm assuming it's not.

File: interfaces/apparmor/spec_test.go [hunk 4]
```diff
@@ -280,3 +280,30 @@ func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
 	c.Assert(updateNS[3], Equals, profile3)
 	c.Assert(updateNS, DeepEquals, []string{profile0, profile1, profile2, profile3})
 }
+
+func (s *specSuite) TestChopTree(c *C) {
```
Comment (chipaca) [inline interfaces/apparmor/spec_test.go]: can we also have tests for non-absolute paths, tests for paths with double slashes, tests for the empty path, tests for negative `assumedPrefixDepth`, and tests for a deeper `assumedPrefixDepth` than the path provided?

File: interfaces/apparmor/spec.go (hunk 3 above)
Comment (zyga) [reply interfaces/apparmor/spec.go]: It's hard to say, we will make a few calls like that per layout item. Not 10Ks maybe ~20 per snap that uses layouts moderately?

File: interfaces/apparmor/spec_test.go (hunk 4 above)
Comment (zyga) [reply interfaces/apparmor/spec_test.go]: Yes, thought the code does run on top of cleaned paths so I could perhaps make it clean the path once again, to make sure it's sane.

**Comment by bboozzoo** [comment]
This branch currently conflicts with master.

File: interfaces/apparmor/spec.go [hunk 5]
```diff
@@ -173,6 +173,96 @@ func isProbablyPresent(path string) bool {
...
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz) assuming that part of that directory already exists (e.g.
+// /foo/) but may need to be created earlier (e.g. in /foo/bar/) because
+// /foo/bar/ is a read-only file system.
+//
+// The mimic works by mounting a tmpfs over the mimicked directory and then
+// re-creating empty files and directories as mount points for the subsequent
+// bind- mount operations to latch onto. This is the reason for the {*,*/,..}
+// part of the expression (we need it to make the empty placeholders).
+func chopTree(path string, assumedPrefixDepth int) (string, string) {
```
Comment (bboozzoo) [inline interfaces/apparmor/spec.go]: Non iterative alternative https://play.golang.org/p/nsqzu1JuxWB

Comment (zyga) [reply interfaces/apparmor/spec.go]: Does it pass tests? :-)

Comment (bboozzoo) [reply interfaces/apparmor/spec.go]: Yeah, I used these test cases https://github.com/snapcore/snapd/pull/5081/files#diff-16cc40a4b635e1e0e4c9d520342fd500R291 Just to be clear, not a blocker or anything, it was just fun to write.

**Comment by zyga** [comment]
I will address the feedback today.

File: interfaces/apparmor/spec.go [hunk 6]
```diff
@@ -173,6 +173,96 @@ func isProbablyPresent(path string) bool {
...
+// expression describes all the files and directories at that level. Coupled
+// with the string "rw ," it can be used to create a rule that allows write
+// access to any file therein, but not deeper.
+//
+// For example, with path: "/foo/bar/froz/baz" and depth 2 the result is:
+// "/{,foo/{,bar/}}" and "/foo/bar/{*,*/,froz/{*,*/,baz}}". Coupled with the
+// aforementioned constants this translates to the following apparmor rules:
+//
+//   /{,foo/{,bar/}} r,
+//   /foo/bar/{*,*/,froz/{*,*/,baz}} rw,
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply interfaces/apparmor/spec.go]: I've pushed a change that implements this right here https://github.com/bboozzoo/snapd/commit/cd38f100676a68b8e3a57d75d5ac0268b05a4f91 @zyga feel free to pick up the patch. The behavior of around edge cases, where split position is > path elements is slightly fishy. Not knowing chopTree will be used makes it non obvious how the original code was planned to be used.

**Comment by zyga** [comment]
Thank you @bboozzoo, I will build upon your patch to get this right. What I think is currently missing is nested expansion of the rules that the original still did, we need the `*` part at various levels, depending on the depth of the path.

**Comment by zyga** [comment]
I will finally get to this today, sorry for the lag. Some prerequisites and helper branches are now in place.

**Comment by zyga** [comment]
I'm re-working this based on prior feedback, more experiments and the new path iterator. I will re-open this shortly.

**Comment by zyga** [comment]
I fixed this now, I just need to get #5319 merged.

File: interfaces/apparmor/spec.go [hunk 7]
```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
...
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz/baz) assuming that part of that directory already exists (e.g.
+// /foo/bar/) but may need to be created earlier (e.g. in /foo/bar/froz).
+//
+// The mimic works by mounting a tmpfs over the mimicked directory and then
+// re-creating empty files and directories as mount points for the subsequent
+// bind- mount operations to latch onto. This is why the right list of
+// expressions use * and */, this allows the expressions to capture files and
+// directories at a specific path.
+func chopTree(path string, assumedPrefixDepth int) ([]string, []string, error) {
```
Comment (bboozzoo) [inline interfaces/apparmor/spec.go]: You can use named return arguments: `func chopTree(...) (left []string, right []string, err error) {`

File: interfaces/apparmor/spec.go [hunk 8]
```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
...
+//
+// The right list describes the remaining directories and the leaf entry
+// (either file or directory) and is somewhat subtle. At each depth level the
+// expression describes all the files and directories at that level. Coupled
+// with the string "rw ," it can be used to create a rule that allows write
+// access to any file therein, but not deeper.
+//
+// For example, with path: "/foo/bar/froz/baz" and depth 3 the result is:
+// []string{"/", "/foo/", "/foo/bar/"} and []string{"/foo/*", "/foo/bar/*",
+// "/foo/bar/froz/*"}. Coupled with the aforementioned constants this
```
Comment (jdstrand) [inline interfaces/apparmor/spec.go]: This comment doesn't coincide with the below rules. With the generated rules below I expected the return value to be: `[]string{"/", "/foo/", "/foo/bar/"}` and `[]string{"/foo/bar/*", "/foo/bar/*/", "/foo/bar/froz/*", "/foo/bar/froz/*/"}` (ie, the right hand side (RHS) in your comment has `"/foo/*"`, but that isn't represented in the rules below). I see in the testsuite updates below that you use exactly what I expected, so this is only a comment update.

File: interfaces/apparmor/spec.go [hunk 9]
```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
...
+//   /foo/bar/froz/*/ rw,
+//
+// Those rules are useful for constructing the apparmor profile for a writable
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz/baz) assuming that part of that directory already exists (e.g.
+// /foo/bar/) but may need to be created earlier (e.g. in /foo/bar/froz).
+//
+// The mimic works by mounting a tmpfs over the mimicked directory and then
+// re-creating empty files and directories as mount points for the subsequent
+// bind- mount operations to latch onto. This is why the right list of
```
Comment (jdstrand) [inline interfaces/apparmor/spec.go]: s/bind- mount/bind mount/

File: interfaces/apparmor/spec.go [hunk 10]
```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
...
+// expressions use * and */, this allows the expressions to capture files and
+// directories at a specific path.
```
Comment (jdstrand) [inline interfaces/apparmor/spec.go]: Not a blocker: you should have enough information to differentiate with the rule (ie, only use * with a file and only use */ with a dir). I'm not sure how much this would complicate the code. It also isn't terribly interesting to make that distinction in the snap-update-ns profile.

File: interfaces/apparmor/spec.go [hunk 11]
```diff
@@ -173,6 +174,92 @@ func isProbablyPresent(path string) bool {
...
+				left = append(left, iter.CurrentPath()+"/")
+			} else {
+				left = append(left, iter.CurrentPath())
+			}
+		} else {
+			// The right hand side replaces the final component with a "*" and
+			// "*/", meaning any file and any directory, respectively.
+			if iter.Depth() > 1 {
+				// The right hand side rules should not allow creation of the
+				// root directory as that itself is meaningless.
```
Comment (jdstrand) [inline interfaces/apparmor/spec.go]: This comment should be moved up as it is talking about '> 1' and not the append operations.

File: interfaces/apparmor/spec_test.go [hunk 12]
```diff
@@ -280,3 +280,75 @@ func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
...
+	}{
+		// Test case from the documentation of the function.
+		{p: "/foo/bar/froz/baz/", d: 3, // Assume first three directories exist
+			l: []string{"/", "/foo/", "/foo/bar/"},
+			// Assume that /foo/bar/froz and beyond may be missing
+			r: []string{"/foo/bar/*", "/foo/bar/*/", "/foo/bar/froz/*", "/foo/bar/froz/*/"}},
+
+		// Exhaustive test cases for directory paths.
+
+		{p: "/foo/bar/froz/", d: 0, // Assume that no directories exist
```
Comment (jdstrand) [inline interfaces/apparmor/spec_test.go]: Perhaps: `// Assume that no directories exist (and '/' does not have 'w')`

File: interfaces/apparmor/spec_test.go [hunk 13]
```diff
@@ -280,3 +280,75 @@ func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
...
+			r: []string{"/*", "/*/", "/foo/*", "/foo/*/", "/foo/bar/*", "/foo/bar/*/"}},
```
Comment (jdstrand) [inline interfaces/apparmor/spec_test.go]: Below with files, you check for `l: []string(nil),` but you have no 'l' check here.

File: interfaces/apparmor/spec_test.go [hunk 14]
```diff
@@ -280,3 +280,75 @@ func (s *specSuite) TestApparmorSnippetsFromLayout(c *C) {
...
+
+		// Assumed prefix depth larger than actual path depth is harmless.
+		{p: "/foo/bar/froz/", d: 5,
+			l: []string{"/", "/foo/", "/foo/bar/", "/foo/bar/froz/"}},
+		{p: "/foo/bar/froz", d: 5,
+			l: []string{"/", "/foo/", "/foo/bar/", "/foo/bar/froz"}},
+
+		// Unclean paths are not allowed.
+		{p: "/foo/../bar", d: 1, e: "cannot chop unclean path: .*"},
+		{p: "/foo//bar", d: 1, e: "cannot chop unclean path: .*"},
```
Comment (jdstrand) [inline interfaces/apparmor/spec_test.go]: Add relative check `{p: "foo/bar", d: 1, ...`?

File: interfaces/apparmor/spec.go (hunk 10 above)
Comment (zyga) [reply interfaces/apparmor/spec.go]: Are you sure? I thought about this as well but the reason for * and */ together is that we need mimic the directory at that specific level. This implies that we need to create empty files and empty directories to re-create the directory as it was there. It doesn't matter if the original path was for a file or for a directory.

File: interfaces/apparmor/spec_test.go (hunk 13 above)
Comment (zyga) [reply interfaces/apparmor/spec_test.go]: l: []string(nil) is the default as it is a golang zero value, should just remove it elsewhere to be consistent

File: interfaces/apparmor/spec.go (hunk 10 above)
Comment (jdstrand) [reply interfaces/apparmor/spec.go]: My point was that looking at this in isolation, if `{p: "/foo/bar/froz/baz/", d: 3`, then the lhs is clearly `{"/", "/foo/", "/foo/bar/"}` but the rhs could be `{"/foo/bar/*", "/foo/bar/*/", "/foo/bar/froz/*/"}`. However, in typing that, I remembered that part of the mimicking behavior outside of chopTree will be to mimic the files in /foo/bar/froz/* *and* `/baz/` since `froz` is being mimicked. The fact that I forgot about that indicates that perhaps the comment should be more clear that the caller of chopTree() expects both the file and directory rules in place to do its job.

File: interfaces/apparmor/spec_test.go (hunk 14 above)
Comment (zyga) [reply interfaces/apparmor/spec_test.go]: This is done, the comment shows because I added it below.

File: interfaces/apparmor/spec.go (hunk 8 above)
Comment (zyga) [reply interfaces/apparmor/spec.go]: This is done
